<a href="https://colab.research.google.com/github/mitchellmui17/CSc-44800/blob/main/Assignment03/Assignment3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


In [89]:
# Imports section
import numpy as np
import pandas as pd
from sklearn import datasets, linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import Perceptron

## Part 1. Loading the dataset

In [5]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
slime = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
slime.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [6]:
slime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

features = slime[["Temperature °C" ,"Mols KCL"]]
features

Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

In [11]:
label = slime[["Size nm^3"]]
label

Size nm^3
0    6.244743e+05
1    5.779610e+05
2    6.196847e+05
3    1.460449e+06
4    4.325726e+04
..            ...
995  1.545661e+06
996  6.737041e+05
997  3.477543e+05
998  8.684794e+05
999  8.476413e+05

[1000 rows x 1 columns]

In [21]:
#specify parameter size.
features_train, features_test, label_train, label_test = train_test_split(features, label, test_size = 0.1)

In [22]:
print(features_train.shape)
print(features_test.shape)
print(label_train.shape)
print(label_test.shape)

(900, 2)
(100, 2)
(900, 1)
(100, 1)


## Part 3. Perform a Linear Regression

In [7]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX


Sample equation: $E = mc^2$

In [30]:
reg = LinearRegression().fit(features_train, label_train)
reg.predict(features_test)

array([[-246581.16144965],
       [ -96567.12008271],
       [ 762339.68994699],
       [ 760888.69080545],
       [ 302666.8559741 ],
       [ 955099.03020294],
       [ 465588.33943868],
       [ 473018.06341259],
       [ 734709.79295428],
       [ 341725.57167351],
       [-251565.97495944],
       [1148995.63941217],
       [ 172968.63055641],
       [ 451165.75738174],
       [1166293.81414788],
       [ 832596.71515725],
       [ 585875.81123607],
       [ 198473.62140037],
       [-117602.16147564],
       [ 945153.28104666],
       [-279753.52670638],
       [ -10250.45579867],
       [ 938446.52890251],
       [ 930427.68908804],
       [1336782.10508998],
       [1042434.7927688 ],
       [ 411120.01380939],
       [ 897158.50328215],
       [ 614475.13236552],
       [ 530411.0465432 ],
       [1063575.456579  ],
       [ 573080.87500534],
       [ 141481.77781654],
       [ 577699.14711823],
       [ 975208.04747713],
       [ 182520.2140664 ],
       [ 116457.75265487],
 

In [35]:
round(reg.score(features_test, label_test), 5)

0.88433

The score shown above, 0.88433 denotes how many points will likely fall on the linear regression line in the test data. To be more specific, it means there will be roughly 88% of the points shown above that will reach the regression line. 

In [38]:
Y_intercept = reg.intercept_
CoE = reg.coef_
print("Intercept: ", Y_intercept)
print("CoEfficient: ", CoE)

Intercept:  [-409188.30540342]
CoEfficient:  [[ 860.66465591 1031.03721347]]


$$H(x) = 860.66465591x - 409188.30540342$$
$$1031.03721347 = 860.66465591x - 409188.30540342$$
$$410219.3426 = 860.66465591x$$
$$x = 476.6309223$$

## Part 4. Use Cross Validation

In [4]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance


In [44]:
cross_val_score(LinearRegression(), features_train, label_train, cv= 7)

array([0.89150523, 0.82878772, 0.8198016 , 0.86131549, 0.85002444,
       0.86540412, 0.84653446])

Based on the information in the array provided, unlike the train and test data we chose earlier, this one I believe as shown switches up the data provided prior above and mixes it. The r values provided in the array show the different percentages of how likely they are to be on the graph based on grabbing different values to be tests. This way it can show more accurate results. If we concluded based on our first regression, we would'nt gotten a lower regression shown in array[2] in the output for part 4. 

## Part 5. Using Polynomial Regression

In [50]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.

poly = PolynomialFeatures(degree=2)
poly.fit_transform(features_train)

array([[1.00000e+00, 9.44000e+02, 9.67000e+02, 8.91136e+05, 9.12848e+05,
        9.35089e+05],
       [1.00000e+00, 2.81000e+02, 7.75000e+02, 7.89610e+04, 2.17775e+05,
        6.00625e+05],
       [1.00000e+00, 4.35000e+02, 9.29000e+02, 1.89225e+05, 4.04115e+05,
        8.63041e+05],
       ...,
       [1.00000e+00, 3.70000e+02, 8.83000e+02, 1.36900e+05, 3.26710e+05,
        7.79689e+05],
       [1.00000e+00, 4.55000e+02, 6.17000e+02, 2.07025e+05, 2.80735e+05,
        3.80689e+05],
       [1.00000e+00, 7.04000e+02, 6.10000e+01, 4.95616e+05, 4.29440e+04,
        3.72100e+03]])

In [65]:
poly_features = poly.fit_transform(features_train)
poly_features

array([[1.00000e+00, 9.44000e+02, 9.67000e+02, 8.91136e+05, 9.12848e+05,
        9.35089e+05],
       [1.00000e+00, 2.81000e+02, 7.75000e+02, 7.89610e+04, 2.17775e+05,
        6.00625e+05],
       [1.00000e+00, 4.35000e+02, 9.29000e+02, 1.89225e+05, 4.04115e+05,
        8.63041e+05],
       ...,
       [1.00000e+00, 3.70000e+02, 8.83000e+02, 1.36900e+05, 3.26710e+05,
        7.79689e+05],
       [1.00000e+00, 4.55000e+02, 6.17000e+02, 2.07025e+05, 2.80735e+05,
        3.80689e+05],
       [1.00000e+00, 7.04000e+02, 6.10000e+01, 4.95616e+05, 4.29440e+04,
        3.72100e+03]])

In [66]:
poly_reg_model = LinearRegression()
poly_reg_model.fit(poly_features, label_train)

LinearRegression()

In [71]:
#Predict the responses 
poly_reg_model.predict(poly_features)

array([[1.86374083e+06],
       [4.56082714e+05],
       [8.38108314e+05],
       [4.31400600e+05],
       [3.31453714e+05],
       [3.92907457e+05],
       [2.00457257e+05],
       [9.62502572e+04],
       [5.40877829e+05],
       [1.03489114e+05],
       [5.94954286e+04],
       [1.47449711e+06],
       [2.77241143e+04],
       [1.86544286e+04],
       [1.15344643e+06],
       [1.50911211e+06],
       [2.59291143e+04],
       [3.74084457e+05],
       [1.93085257e+05],
       [1.07004960e+06],
       [4.98928286e+04],
       [1.53596811e+06],
       [1.96465029e+05],
       [1.04813600e+06],
       [1.46044903e+06],
       [3.50058714e+05],
       [7.74480114e+05],
       [5.10753257e+05],
       [2.39246314e+05],
       [6.13645600e+05],
       [2.42666829e+05],
       [7.24974257e+05],
       [2.74972029e+05],
       [4.07851114e+05],
       [1.07782503e+06],
       [3.35550429e+05],
       [1.19722029e+05],
       [5.89426857e+05],
       [4.70630400e+05],
       [3.76202257e+05],


In [73]:
poly_reg_model.score(poly_features, label_train)

1.0

In [74]:
poly_reg_model.coef_


array([[ 0.00000000e+00,  1.20000000e+01, -9.57353624e-08,
         1.33226763e-11,  2.00000000e+00,  2.85714287e-02]])

In [75]:
poly_reg_model.intercept_

array([1.41797937e-05])

It is quite surprising, but the data is at a 1.0, I wasn't expecting a 100 percent accuracy. It improved the prior models by a minimum of ~15 percent increase from the prior models. 

$$\hat{y}(w,x) =  0 + 12w - 0x + 0w^2 + 2wx + 0.02857x^2 - 0$$